pkg/billing/model.go

```diff
@@ -23,3 +26,44 @@ type IncrementalEvent struct {
 	StopTime       time.Time `json:"stop_time"`
 	Value          int       `json:"value"`
 }
+
+type enrichable interface {
+	setEventType()
+	idempotencyKey() *string
+}
+
+// Enrich sets the event's Type and IdempotencyKey fields, so that users of this API don't need to
+// manually set them
+//
+// Enrich is already called by (*Batch).Add, but this is used in the autoscaler-agent's billing
+// implementation, which manually calls Enrich in order to log the IdempotencyKey for each event.
+func Enrich[E enrichable](batch *Batch, event E) E {
+	event.setEventType()
+
+	key := event.idempotencyKey()
+	if *key == "" {
+		*key = fmt.Sprintf("Host<%s>:ID<%s>:T<%s>", batch.c.hostname, uuid.NewString(), time.Now().Format(time.RFC3339))
+	}
+
+	return event
+}
+
+// setEventType implements enrichable
+func (e *AbsoluteEvent) setEventType() {
+	e.Type = "absolute"
+}
+
+// idempotencyKey implements enrichable
+func (e *AbsoluteEvent) idempotencyKey() *string {
+	return &e.IdempotencyKey
+}
+
+// setEventType implements enrichable
+func (e *IncrementalEvent) setEventType() {
+	e.Type = "incremental"
+}
+
+// idempotencyKey implements enrichable
+func (e *IncrementalEvent) idempotencyKey() *string {
+	return &e.IdempotencyKey
+}
```

Comment:
can you group methods in this file by type?

Reply:
Not sure what you mean, can you elaborate?

Reply:
sorry, 
```go
type foo struct{}
func (foo) thingOne() {}
func (foo) thingTwo() {}
type bar struct{}
func (bar) thingOne() {}
func (bar) thingTwo() {}
```